internal/session/cache.go

```diff
@@ -0,0 +1,98 @@
+package session
+
+import (
+	"context"
+	"log/slog"
+	"sync"
+
+	redis "github.com/redis/go-redis/v9"
+)
+
+type Cache struct {
+	connectionString string
+	inmemory         *sync.Map
+	extClient        *redis.Client
+	logger           *slog.Logger
+}
+
+func (c *Cache) KeyExists(ctx context.Context, key string) (bool, error) {
+	if c.inmemory != nil {
+		_, ok := c.inmemory.Load(key)
+		return ok, nil
+	}
+	count, err := c.extClient.Exists(ctx, key).Result()
+	if err != nil {
+		return false, err
+	}
+	if count > 0 {
+		return true, nil
+	}
+	return false, nil
+
+}
+
+func (c *Cache) GetSession(ctx context.Context, key string) (map[string]string, error) {
+	if c.inmemory != nil {
+		val, ok := c.inmemory.Load(key)
+		if ok {
+			return val.(map[string]string), nil
+		}
+		return map[string]string{}, nil
+	}
+	return c.extClient.HGetAll(ctx, key).Result()
+}
+
+func (c *Cache) DeleteSessions(ctx context.Context, key ...string) error {
+	if c.inmemory != nil {
+		for _, k := range key {
+			c.inmemory.Delete(k)
+		}
+		return nil
+	}
+	return c.extClient.Del(ctx, key...).Err()
+}
+
+// AddSession will add a session under the key. If the key exists it will append that session
+func (c *Cache) AddSession(ctx context.Context, key, mcpServerID, mcpSession string) (bool, error) {
+	if c.inmemory != nil {
+		session, err := c.GetSession(ctx, key)
+		if err != nil {
+			return false, err
+		}
+		session[mcpServerID] = mcpSession
+		c.inmemory.Store(key, session)
+		return true, nil
+	}
+	err := c.extClient.HSet(ctx, key, mcpServerID, mcpSession).Err()
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func NewCache(logger *slog.Logger, opts ...func(*Cache)) (*Cache, error) {
+	c := &Cache{
+		inmemory: nil,
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	if c.connectionString != "" {
+		opt, err := redis.ParseURL(c.connectionString)
+		if err != nil {
+			return c, err
+		}
+		c.extClient = redis.NewClient(opt)
```

Comment:
reconnects handled automatically?

Reply:
I think it is likely we might want to set some other settings but at the moment this will take a credential from the connection string and connect using that. We could do a ping as I don't think the new connection will error if not available